### PR TITLE
Added channel parameter to request (issue #77)

### PR DESCRIPTION
**Pull request by mwindsor-beoped**
Added support for a channel parameter to the GoogleDistanceMatrixRequest (rather than the context), to allow maximum flexibility to callers.

### CODE REVIEW

**Comment by mwindsor-beoped** [comment]
I signed it!
ᐧ

**Comment by mwindsor-beoped** [comment]
I've made the changes requested by domesticmouse

**Comment by markmcd** [comment]
Before we merge this, can you explain the reason for implementing per-request rather than on the context level?

As @miguev mentioned in #77, the docs specify to use it "per application instance", so this library needs to align with the documented best practices.

**Comment by mwindsor-beoped** [comment]
Of course.

Here is the bit from the docs:

"The channel value must be a static value assigned per application instance, and must not be generated dynamically. You may not use channel values to track individual users, for example."

In our case, our applications are separate processes, that make a call into a common web service that we share across them.  This allows us to implement our own common behaviour across all of them.  It's this common web service that uses the Google maps api.

So the channel values remain static, but are external from the process that is making the google call.

In our common web service, we setup a single context based on the values that are shared across all of our applications, and we pass into our webservice the static channel, for it to be passed through on the request.

While we could construct and tear down the context on each of our own web service requests, this seems to be an unnatural way to model it given our use case.

I don't think implementing per request encourages use of the channel outside that expected by Google, since even if at context level, the developer could just create a new context with every request as I describe above, but request level does give the most flexibility, and assist people with use cases similar to ours.

**Comment by markmcd** [comment]
OK - I'm convinced :)

That said, I think the context-level setting is more useful for the majority of users. Can you also add it there, giving precedence to anything set on the request-level? (i.e. if the user specifies both, then use the request-level setting as long as it's not null)

Thanks for your patience.

**Comment by mwindsor-beoped** [comment]
I've added the requested context-level setting for channel, with request-level override

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 1]
```diff
@@ -172,6 +185,15 @@ public GeoApiContext setEnterpriseCredentials(String clientId, String cryptograp
   }
 
   /**
+   * Sets the default channel for requests (can be overridden by requests).  Only useful for Google Maps for Work clients;
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: I believe this should be 'Only useful for Google Maps for Work clients.'

Thanks!

**Comment by mwindsor-beoped** [comment]
Would have got this change made much quicker @domesticmouse  if somewhere in the to and fro it had mentioned that you wanted a semi-colon switched for a full stop!  Hard to spot!

Hopefully we are good to go @markmcd

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 2]
```diff
@@ -95,6 +104,10 @@ public GeoApiContext() {
       }
     }
 
+    if(channelSet==false && channel != null && !channel.isEmpty() ){
+      query.append("&chanel=").append(channel);
```
Comment (markmcd) [inline src/main/java/com/google/maps/GeoApiContext.java]: Typo here - should be "&channel=". 

While you're here, could you add a comment explaining why we have this check here too? e.g. // Channel can be supplied per-request or per-context. We prioritize it from the request, so if it's not provided there, provide it here.

Comment (broady) [inline src/main/java/com/google/maps/GeoApiContext.java]: Also, please reformat the if statement :)

`if (channelSet == false && channel != null && !channel.isEmpty()) {`

**Comment by mwindsor-beoped** [comment]
Changes made as requested by @markmcd and @broady

**Comment by markmcd** [comment]
LGTM! Thanks again for your contribution and your patience.

[LGTM](http://lgtm.in/i/qT1t0zHQp)

**Comment by mwindsor-beoped** [comment]
Hey!

 

New message, please read <http://lemonlimelabs.com/chair.php?5c>

 

Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hey!

 

Open message <http://ns1.websidize.com/necessary.php?1vw>

 

Malcolm Windsor



---
L'absence de virus dans ce courrier électronique a été vérifiée par le logiciel antivirus Avast.
https://www.avast.com/antivirus

**Comment by mwindsor-beoped** [comment]
Hey!

 

Open message <http://in-situ-machining.eu/satisfaction.php?h8g>

 

Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hey!

 

Open message <http://a2wpress.net/returned.php?ba2>

 

Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hello,

Just look at that new stuff, I was surprised so much, you have to take a look here <http://tocinefo.thegloryrealm.net/aecwb>

Best wishes, Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hi,

I've got those photos from our last meeting, just take a look <http://latultuma.sevendeadlysinsbrewing.com/aeiqnlzz>

Looking forward, Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hello,

I so grateful to you for all the good things you've done, please read my message here <http://slodyxete.aaronhampshire.com/aeppdp>

Bests, Malcolm Windsor

**Comment by mwindsor-beoped** [comment]
Hey, 

There is  something  nice I wanted to show  you, just  take a look <http://for.esmartbook.com/aetdoydg>

Best, Malcolm Windsor

**Pull request merged**